src/geode/mesh/helpers/euclidean_distance_transform.cpp

```diff
@@ -0,0 +1,392 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+#include <geode/mesh/helpers/euclidean_distance_transform.h>
+
+#include <async++.h>
+
+#include <geode/basic/attribute_manager.h>
+#include <geode/basic/progress_logger.h>
+
+#include <geode/mesh/core/regular_grid_solid.h>
+#include <geode/mesh/core/regular_grid_surface.h>
+
+namespace geode
+{
+    template < index_t dimension >
+    class EuclideanDistanceTransform
+    {
+        using Index = GridCellIndices< dimension >;
+
+    public:
+        EuclideanDistanceTransform( const RegularGrid< dimension >& grid,
+            absl::Span< const GridCellIndices< dimension > > grid_cell_id,
+            absl::string_view distance_map_name )
+            : grid_{ grid }
+        {
+            distance_map_ =
+                grid_.cell_attribute_manager()
+                    .template find_or_create_attribute< VariableAttribute,
+                        double >( distance_map_name,
+                        std::numeric_limits< double >::max() );
+
+            for( const auto& cell_id : grid_cell_id )
+            {
+                distance_map_->set_value( grid_.cell_index( cell_id ), 0. );
+            }
+        }
+
+        std::shared_ptr< VariableAttribute< double > > distance_map() const
+        {
+            return distance_map_;
+        }
+
+        void compute_distance_map();
+        void directional_squared_distance( const index_t direction );
+        void combine_squared_distance_components( const index_t direction );
+
+        void squared_root_filter()
+        {
+            async::parallel_for(
+                async::irange( index_t{ 0 }, grid_.nb_cells() ),
+                [this]( index_t cell ) {
+                    distance_map_->modify_value( cell, []( double& value ) {
+                        value = std::sqrt( value );
+                    } );
+                } );
+            return;
+        }
+
+    private:
+        double directional_step_squared_distance( const Index& from_index,
+            const Index& to_index,
+            const index_t direction,
+            const double last_step_squared_distance )
+        {
+            const auto squared_cell_length =
+                grid_.cell_length_in_direction( direction )
+                * grid_.cell_length_in_direction( direction );
```

Comment:
these could be computed once in the constructor and stored in an array as member